cs/src/Management/TunnelManagementClient.cs

```diff
@@ -924,6 +925,60 @@ public void Dispose()
             CancellationToken cancellation)
         {
             Requires.NotNull(tunnel, nameof(tunnel));
+            options ??= new TunnelRequestOptions();
+            options.AdditionalHeaders ??= new List<KeyValuePair<string,string>>();
+            options.AdditionalHeaders = options.AdditionalHeaders.Append(new KeyValuePair<string, string>("If-None-Match", "*"));
+            var tunnelId = tunnel.TunnelId;
+            var idGenerated = string.IsNullOrEmpty(tunnelId);
+            if (idGenerated)
+            {
+                tunnel.TunnelId = IdGeneration.GenerateTunnelId();
+            }
+            for (int retries = 0; retries <= 3; retries++)
+            {
+                try
+                {
+                    var result = await this.SendTunnelRequestAsync<Tunnel, Tunnel>(
+                       HttpMethod.Put,
+                       tunnel,
+                       ManageAccessTokenScope,
+                       path: null,
+                       query: GetApiQuery(),
+                       options,
+                       ConvertTunnelForRequest(tunnel),
+                       cancellation,
+                       true);
+                    PreserveAccessTokens(tunnel, result);
+                    return result!;
+                }
+                catch (UnauthorizedAccessException) when (idGenerated && retries < 3) // The tunnel ID was already taken.
+                {
+                    tunnel.TunnelId = IdGeneration.GenerateTunnelId();
+                }
```

Comment:
We could consider catching an "expectation failed" exception when `idGenerated == false` to skip additional retries.

Reply:
Since this is a "catch when" if idgenerated == false the exception will not be caught and no retries will be done

Reply:
I was thinking about a totally separate branch - basically, if the ID is specified and the creation fails because the ID is taken, it's not worth retrying. It's not an essential piece of logic, and could definitely be optimized later.

Reply:
That should be the current behavior. In that case the catch would not happen so the Exception would just be thrown and that would be it

Reply:
Ah, you're right! Thanks!